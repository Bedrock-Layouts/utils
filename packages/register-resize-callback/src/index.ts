export type ResizeFunc = (r: ResizeObserverEntry) => unknown;
export type DeregisterFunc = () => unknown | Error;

type ObserverCallback =
  | ResizeFunc
  | {
      current: ResizeFunc;
    };

// eslint-disable-next-line functional/no-let
let observer: ResizeObserver | undefined;

const callBackMap = new WeakMap<Element, ObserverCallback[]>();

export function init(): ResizeObserver {
  /* c8 ignore next */
  if (observer) return observer;

  return (observer = new ResizeObserver((entries) => {
    return entries.map((entry) => {
      const maybeCallBacks = callBackMap.get(entry.target);
      /**
       * This is defensive code, but it's possible that get returns undefined
       */
      /* c8 ignore next */
      const safeCallBacks = maybeCallBacks ?? [];

      if (safeCallBacks.length === 0) return entry;

      requestAnimationFrame(() => {
        return safeCallBacks.map((callback) => {
          const safeCallback =
            typeof callback === "function" ? callback : callback.current;

          safeCallback(entry);

          return callback;
        });
      });
      return entry;
    });
  }));
}

export function registerCallback(
  node: Readonly<Element>,
  // eslint-disable-next-line functional/prefer-immutable-types
  cb: ObserverCallback,
): DeregisterFunc {
  if (observer === undefined) {
    init();
  }

  const callbacks = callBackMap.get(node) ?? [];

  callBackMap.set(node, callbacks.concat(cb));
  observer?.observe(node);

  return () => {
    const maybeCallBacks = callBackMap.get(node);
    /**
     * This is defensive code, but it's possible that get returns undefined
     */
    /* c8 ignore next */
    const safeCallBacks = maybeCallBacks ?? [];
    const newCallbacks = safeCallBacks.filter((ref) => ref !== cb);

    callBackMap.set(node, newCallbacks);

    if (newCallbacks.length === 0) {
      observer?.unobserve(node);
      callBackMap.delete(node);
    }
  };
}
