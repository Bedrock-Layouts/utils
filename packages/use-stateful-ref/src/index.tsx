import React from "react";

export function useStatefulRef<T>(initialVal?: T): React.MutableRefObject<T> {
  // eslint-disable-next-line prefer-const, functional/no-let
  let [cur, setCur] = React.useState<T | undefined>(initialVal);

  const { current: ref } = React.useRef({
    current: cur,
  });

  // eslint-disable-next-line functional/immutable-data
  Object.defineProperty(ref, "current", {
    get: () => cur as T,
    // eslint-disable-next-line functional/no-return-void
    set: (value: T) => {
      if (!Object.is(cur, value)) {
        cur = value;
        setCur(value);
      }
    },
  });

  return ref as React.MutableRefObject<T>;
}
