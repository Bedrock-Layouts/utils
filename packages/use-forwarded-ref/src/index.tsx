import { useStatefulRef } from "@bedrock-layout/use-stateful-ref";
import { useRef } from "react";
import React from "react";

export interface Config {
  isStateful: boolean;
}

export function useForwardedRef<T>(
  forwardedRef?: React.Ref<T>,
  config: Config = { isStateful: true },
): React.MutableRefObject<T> {
  const statefulRef = useStatefulRef<T>();
  const ref = useRef<T>();

  const innerRef = config.isStateful ? statefulRef : ref;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  React.useImperativeHandle(forwardedRef, () => innerRef.current);

  return innerRef as React.MutableRefObject<T>;
}
