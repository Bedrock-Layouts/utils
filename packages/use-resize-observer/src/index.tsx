import {
  ResizeFunc,
  init,
  registerCallback,
} from "@bedrock-layout/register-resize-callback";
import { useStatefulRef } from "@bedrock-layout/use-stateful-ref";
import { useEffect, useLayoutEffect, useRef } from "react";

export function useResizeObserver<T extends Element>(
  callback: ResizeFunc,
  node?: T,
): React.MutableRefObject<T> {
  const callbackRef = useRef<ResizeFunc>(callback);

  const nodeRef = useStatefulRef<T>();

  // eslint-disable-next-line functional/no-return-void
  useLayoutEffect(() => {
    // eslint-disable-next-line functional/immutable-data
    callbackRef.current = callback;
  });

  // eslint-disable-next-line functional/no-return-void
  useEffect(() => {
    init();
  }, []);

  const currentNode = nodeRef.current ?? node;

  useEffect(() => {
    init();

    /**
     * node is undefined when the component is unmounted or not yet mounted
     */
    /* c8 ignore start */
    const cleanup = currentNode
      ? registerCallback(currentNode, callbackRef)
      : () => 0;
    /* c8 ignore end */

    // eslint-disable-next-line functional/no-return-void
    return () => {
      cleanup();
    };
  }, [currentNode]);

  return nodeRef;
}
