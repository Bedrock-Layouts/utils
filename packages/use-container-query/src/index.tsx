import { useForwardedRef } from "@bedrock-layout/use-forwarded-ref";
import { useResizeObserver } from "@bedrock-layout/use-resize-observer";
import React from "react";

interface UseContainerQueryProps {
  width?: number;
  maxWidth?: number;
}

export function useContainerQuery<T extends Element>(
  { width = 1, maxWidth }: Readonly<UseContainerQueryProps>,
  forwardedRef?: React.Ref<T>,
): [boolean, React.MutableRefObject<T>] {
  const [matches, setMatch] = React.useState(false);
  const containerRef = useForwardedRef(forwardedRef);

  useResizeObserver((entry: ResizeObserverEntry) => {
    //fix typings
    /* c8 ignore start */
    const nodeWidth =
      (entry.borderBoxSize as unknown as ResizeObserverSize)?.inlineSize ??
      entry.contentRect.width;
    /* c8 ignore end */
    //nodeWidth can be zero when it is switching from one node to another.  This will ignore that.
    if (nodeWidth > 0) {
      const newMatch =
        maxWidth === undefined
          ? nodeWidth <= width
          : nodeWidth >= width && nodeWidth <= maxWidth;

      setMatch(newMatch);
    }

    return entry;
  }, containerRef.current);

  return [matches, containerRef];
}
