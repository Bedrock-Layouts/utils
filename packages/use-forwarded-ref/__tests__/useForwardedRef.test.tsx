import { useStatefulRef } from "@bedrock-layout/use-stateful-ref";
import * as React from "react";
import ReactDOM from "react-dom/client";
import { act } from "react-dom/test-utils";
import {
  Mock,
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
  test,
  vi,
} from "vitest";

import * as imports from "../src";

const { useForwardedRef } = imports;

vi.mock("@bedrock-layout/use-stateful-ref", () => {
  const fn = vi.fn(() => {
    let value: unknown = undefined;
    return {
      set current(val) {
        value = val;
      },
      get current() {
        return value;
      },
    };
  });
  return {
    default: fn,
    useStatefulRef: fn,
  };
});

let safeRef: React.MutableRefObject<unknown> | undefined;
const HookWrapper = React.forwardRef(
  ({ isStateful }: { isStateful?: boolean }, ref) => {
    const config = isStateful === undefined ? undefined : { isStateful };
    safeRef = useForwardedRef(ref, config);
    safeRef.current = "safeRef:current";
    return null;
  },
);

describe("useForwardedRef", () => {
  let container: Node;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    safeRef = undefined;

    document.body.removeChild(container);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    container = null;
  });

  test("useForwardedRef is not null", () => {
    expect(useForwardedRef).toBeTruthy();
  });

  test("useStatefulRef is called", () => {
    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(<HookWrapper />);
    });

    expect(useStatefulRef).toBeCalled();
    expect(safeRef).toMatchSnapshot();
  });

  it("should call ref callback", () => {
    const ref = vi.fn();
    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(<HookWrapper ref={ref} />);
    });
    expect(ref).toBeCalled();
  });

  it("should update a ref object", () => {
    const ref = { current: undefined };
    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(<HookWrapper ref={ref} />);
    });
    expect(ref).toMatchSnapshot();
  });

  // Test for isStateful config
  it("should use stateful ref when isStateful is true", () => {
    const mockUseStatefulRef = useStatefulRef as Mock;
    mockUseStatefulRef.mockClear();

    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(<HookWrapper isStateful={true} />);
    });

    expect(mockUseStatefulRef).toHaveBeenCalled();
    expect(safeRef?.current).toBe("safeRef:current");
  });

  it("should use regular ref when isStateful is false", () => {
    const mockUseStatefulRef = useStatefulRef as Mock;
    mockUseStatefulRef.mockClear();

    // Create original ref to track if it's used
    const originalRef = { current: "original" };

    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(
        <HookWrapper isStateful={false} ref={originalRef} />,
      );
    });

    // When isStateful is false, we shouldn't be using the stateful ref for storing values
    expect(mockUseStatefulRef).toHaveBeenCalledTimes(1); // Still called as part of the hook
    expect(safeRef?.current).toBe("safeRef:current");
    expect(originalRef.current).toBe("safeRef:current"); // The ref object should be updated
  });

  it("should use stateful ref by default when config is not provided", () => {
    const mockUseStatefulRef = useStatefulRef as Mock;
    mockUseStatefulRef.mockClear();

    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(<HookWrapper />);
    });

    expect(mockUseStatefulRef).toHaveBeenCalled();
    expect(safeRef?.current).toBe("safeRef:current");
  });
});
