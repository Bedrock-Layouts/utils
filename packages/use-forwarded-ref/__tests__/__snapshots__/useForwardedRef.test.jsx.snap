// Vitest Snapshot v1

exports[`useForwardedRef > should update a ref object 1`] = `
{
  "current": "safeRef:current",
}
`;

exports[`useForwardedRef > useStatefulRef is called 1`] = `
{
  "current": "safeRef:current",
}
`;
